plugins/hls-alternate-number-format-plugin/src/Ide/Plugin/AlternateNumberFormat.hs

```diff
@@ -0,0 +1,134 @@
+{-# LANGUAGE TypeOperators              #-}
+
+module Ide.Plugin.AlternateNumberFormat where
+
+import           Control.Lens                         ((^.))
+import           Control.Monad.Except                 (ExceptT, MonadIO, liftIO)
+import qualified Data.HashMap.Strict                  as HashMap
+import           Data.Maybe                           (fromMaybe, isJust)
+import           Data.Text                            (Text)
+import qualified Data.Text                            as T
+import           Debug.Trace                          (traceM)
+import           Development.IDE                      (GetParsedModule (GetParsedModule),
+                                                       IdeState, RuleResult,
+                                                       Rules, define, ideLogger,
+                                                       realSrcSpanToRange,
+                                                       runAction, use,
+                                                       useWithStale)
+import           Development.IDE.Core.PositionMapping (PositionMapping)
+import           Development.IDE.GHC.Compat           hiding (getSrcSpan)
+import           Development.IDE.GHC.Compat.Util      (toList)
+import           Development.IDE.Graph.Classes        (Hashable, NFData)
+import           Development.IDE.Types.Logger         as Logger
+import           GHC.Generics                         (Generic)
+import           Ide.Plugin.Conversion                (FormatType,
+                                                       alternateFormat,
+                                                       toFormatTypes)
+import           Ide.Plugin.Literals                  (Literal (..),
+                                                       collectLiterals,
+                                                       getSrcSpan, getSrcText)
+import           Ide.Plugin.Retrie                    (handleMaybe,
+                                                       handleMaybeM, response)
+import           Ide.Types
+import           Language.LSP.Types
+import           Language.LSP.Types.Lens              (uri)
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor plId = (defaultPluginDescriptor plId)
+    { pluginHandlers = mkPluginHandler STextDocumentCodeAction codeActionHandler
+    , pluginRules = collectLiteralsRule
+    }
+
+data CollectLiterals = CollectLiterals
+                     deriving (Show, Eq, Generic)
+
+instance Hashable CollectLiterals
+instance NFData CollectLiterals
+
+type instance RuleResult CollectLiterals = CollectLiteralsResult
+
+data CollectLiteralsResult = CLR {
+    literals      :: [Literal]
+    , formatTypes :: [FormatType]
+    } deriving (Generic)
+
+instance Show CollectLiteralsResult where
+    show _ = "<CollectLiteralResult>"
+
+instance NFData CollectLiteralsResult
+
+collectLiteralsRule :: Rules ()
+collectLiteralsRule = define $ \CollectLiterals nfp -> do
```

Comment:
What's the motivation for defining this rule over a plain function?

Reply:
To be quite honest, prior to my knowledge about HLS internals I thought I needed a Diagnostic for each Code Action. The only way I found was via creating the Rule. I'm happy to translate back to a plain function.

Reply:
I think the motivation is to avoid running the expensive AST traversal more often than strictly necessary. Rule results get cached, so the `CollectLiterals` result will be reused if the parsed module hasn't changed. On the other hand, without the rule, the AST traversal will be evaluated every time that the code action handler is invoked, right? So you probably do want the rule.

Reply:
Ahh! Sorry I misinterpreted your first comment. I can revert the changes I made. Would you prefer if I used any other `use` combinator? I currently just used `use` but I can see maybe `useWithStaleFast` being more appropriate?

Reply:
You should validate my conjecture as I am not quite sure of whether the client calls the code action handler more than once. 

Using stale data is optional, I think use is probably fine since a parse tree is almost always available.

Reply:
Based on some rudimentary testing, it looks like the actual traversal portion is called much less when placed in the rule. So unless there are objections, it seems Rule-based is the better approach.